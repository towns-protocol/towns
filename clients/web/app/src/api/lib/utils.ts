import { env, hasGoerliParam, hasVitalkTokensParam } from 'utils'

/**
 * This file is purely for network calls to Alchemy that are using the token-worker
 * These calls are independent of any network the user is connected to, or our app
 * is configured against, or the matrix homeserver.
 *
 * These vars are used in specific points for fetching data from ALchmey's NFT api
 * for prototyping features.
 *
 * TODO: we need to accommodate sepolia
 */

export const fetchVitalikTokens = env.IS_DEV && hasVitalkTokensParam()
export const fetchGoerli = env.IS_DEV && hasGoerliParam()
export const GOERLI = 'eth-goerli'
export const MAINNET = 'eth-mainnet'
export const NETWORK = fetchVitalikTokens ? MAINNET : GOERLI
