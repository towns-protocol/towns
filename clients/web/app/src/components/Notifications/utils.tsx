import React from 'react'
import { PrivyNotAuthenticatedNotification } from '@components/Notifications/PrivyNotAuthenticatedNotification'

import { popupToast } from './popupToast'

export function createPrivyNotAuthenticatedNotification() {
    return popupToast(({ toast }) => {
        return <PrivyNotAuthenticatedNotification toast={toast} />
    })
}
