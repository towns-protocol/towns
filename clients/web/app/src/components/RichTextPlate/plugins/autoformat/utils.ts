import { AutoformatBlockRule } from '@udecode/plate-autoformat'
import { ELEMENT_PARAGRAPH } from '@udecode/plate-paragraph'
import {
    ELEMENT_CODE_BLOCK,
    ELEMENT_CODE_LINE,
    insertEmptyCodeBlock,
} from '@udecode/plate-code-block'
import {
    ELEMENT_DEFAULT,
    PlateEditor,
    getParentNode,
    isElement,
    isType,
} from '@udecode/plate-common'
import { toggleList, unwrapList } from '@udecode/plate-list'

export const preFormat: AutoformatBlockRule['preFormat'] = (editor) => unwrapList(editor)

export const format = (editor: PlateEditor, customFormatting: () => void) => {
    if (editor.selection) {
        const parentEntry = getParentNode(editor, editor.selection)
        if (!parentEntry) {
            return
        }
        const [node] = parentEntry
        if (
            isElement(node) &&
            !isType(editor, node, ELEMENT_CODE_BLOCK) &&
            !isType(editor, node, ELEMENT_CODE_LINE)
        ) {
            customFormatting()
        }
    }
}

export const formatList = (editor: PlateEditor, elementType: string) => {
    format(editor, () =>
        toggleList(editor, {
            type: elementType,
        }),
    )
}

export const formatText = (editor: PlateEditor, text: string) => {
    format(editor, () => editor.insertText(text))
}

export const formatCodeBlock = (editor: PlateEditor) => {
    insertEmptyCodeBlock(editor, {
        defaultType: ELEMENT_DEFAULT,
        insertNodesOptions: { select: true },
    })
}

export const isParagraph = (editor: PlateEditor) => {
    if (!editor.selection) {
        return false
    }

    const parentEntry = getParentNode(editor, editor.selection)
    if (!parentEntry) {
        return false
    }
    const [node] = parentEntry
    return isType(editor, node, ELEMENT_DEFAULT) || isType(editor, node, ELEMENT_PARAGRAPH)
}
