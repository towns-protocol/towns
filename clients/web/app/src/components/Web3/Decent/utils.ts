import { UseBoxActionArgs } from '@decent.xyz/box-hooks'
import { BoxActionResponse, EvmTransaction } from '@decent.xyz/box-common'
import { Address, EstimateGasParameters } from 'viem'

export function isActionConfig(config: unknown): config is UseBoxActionArgs {
    return (
        typeof config === 'object' &&
        config !== null &&
        'sender' in config &&
        typeof config.sender === 'string' &&
        'srcToken' in config &&
        typeof config.srcToken === 'string' &&
        'dstToken' in config &&
        typeof config.dstToken === 'string' &&
        'srcChainId' in config &&
        typeof config.srcChainId === 'number' &&
        'dstChainId' in config &&
        typeof config.dstChainId === 'number' &&
        'slippage' in config &&
        typeof config.slippage === 'number' &&
        'actionType' in config &&
        typeof config.actionType === 'string' &&
        'actionConfig' in config &&
        typeof config.actionConfig === 'object' &&
        config.actionConfig !== null &&
        'amount' in config.actionConfig &&
        typeof config.actionConfig.amount === 'bigint' &&
        'swapDirection' in config.actionConfig &&
        typeof config.actionConfig.swapDirection === 'string' &&
        'receiverAddress' in config.actionConfig &&
        typeof config.actionConfig.receiverAddress === 'string' &&
        'chainId' in config.actionConfig &&
        typeof config.actionConfig.chainId === 'number'
    )
}

export function estimateGasArgs(args: {
    sender: Address | undefined
    boxActionResponse: BoxActionResponse | undefined
}) {
    const { sender, boxActionResponse } = args
    return {
        account: sender,
        ...(boxActionResponse?.tx as EvmTransaction),
    } as unknown as EstimateGasParameters
}

export function isBoxActionResponseError(
    boxActionResponse: BoxActionResponse | BoxActionResponseError | undefined,
): boxActionResponse is BoxActionResponseError {
    if (!boxActionResponse) {
        return false
    }
    return 'error' in boxActionResponse
}

// undocumented and incorrectly typed response from a box action
// can sometimes return this response. try swapping 999 ETH and you get this
type BoxActionResponseError = {
    error: unknown
    success: false
}
