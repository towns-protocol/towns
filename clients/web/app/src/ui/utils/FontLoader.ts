export const FontFamily = {
    TitleFont: 'TitleFont',
    BodyFont: 'BodyFont',
} as const

const AkkuratCapSize = {
    trimTop: '-0.26em',
    trimBottom: '-0.24em',
    lineHeight: '1.24em',
}

const AkkuratStyles = {
    letterSpacing: '-0.02em',
}

export const fontSettings = [
    {
        fontFamily: FontFamily.BodyFont,
        styles: AkkuratStyles,
        src: "url('/fonts/AkkuratLLWeb-Regular.woff2')",
        fontDescription: {
            weight: 'normal',
            style: 'normal',
        },
        targets: ['p', 'ul', 'ol', 'h3', 'h4', 'h5', 'h6'],
        capSize: AkkuratCapSize,
    },
    {
        fontFamily: FontFamily.BodyFont,
        styles: { ...AkkuratStyles },
        src: "url('/fonts/AkkuratLLWeb-Italic.woff2')",
        fontDescription: {
            weight: 'normal',
            style: 'italic',
        },
        targets: [],
        capSize: AkkuratCapSize,
    },
    {
        fontFamily: FontFamily.BodyFont,
        styles: { ...AkkuratStyles },
        src: "url('/fonts/AkkuratLLWeb-Bold.woff2')",
        fontDescription: {
            weight: '600',
            style: 'normal',
        },
        targets: ['p', 'ul', 'ol', 'h3', 'h4', 'h5', 'h6'],
        capSize: AkkuratCapSize,
    },
    {
        fontFamily: FontFamily.TitleFont,
        src: "url('/fonts/MangoGrotesque-Black.ttf')",
        styles: {
            letterSpacing: '0',
        },
        fontDescription: {
            weight: 'bold',
            style: 'normal',
        },
        targets: ['h1', 'h2'],
        capSize: {
            trimTop: '-0.2em',
            trimBottom: '-0.35em',
            lineHeight: '1.24em',
        },
    },
] as const

let isInit = false

export const FontLoader = {
    init() {
        if (isInit) return
        isInit = true
        fontSettings.forEach((f) => {
            const font = new FontFace(f.fontFamily, f.src, f.fontDescription)
            document.fonts.add(font)
        })
    },
}
