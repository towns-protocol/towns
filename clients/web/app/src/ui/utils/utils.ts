import { RoomIdentifier } from 'use-zion-client'
import { PATHS } from 'routes'

export const toPx = (value?: string | number) => (typeof value === 'number' ? `${value}px` : value)

export const notUndefined = <T>(x: T | undefined): x is T => {
    return typeof x !== 'undefined'
}

/**
 * Utility to assign default string values if the incoming value is a bool.
 */
export const assignBoolToDefaultValue = <T>(value: boolean | T, trueValue: T, falseValue?: T) => {
    return typeof value !== 'boolean' ? undefined : value === true ? trueValue : falseValue
}

export const shortAddress = (address: string) => {
    const start = address.slice(0, 5)
    const end = address.slice(-3)
    return `${start}...${end}`
}

export const isRejectionError = (error: Error): boolean => {
    return error.name === 'ACTION_REJECTED'
}

export const isForbiddenError = (error: Error): boolean => {
    return error.name === 'M_FORBIDDEN'
}

export const getInviteUrl = (spaceId: RoomIdentifier | undefined) =>
    `${window.location.protocol}//${window.location.host}/${PATHS.SPACES}/${spaceId?.slug}/?invite`

type Entries<T> = {
    [K in keyof T]: [K, T[K]]
}[keyof T][]

// same as Object.entries but with typed keys
export const getTypedEntries = <T extends object>(obj: T) => Object.entries(obj) as Entries<T>

// Supported name format for channel
export const ChannelNameRegExp = new RegExp(/^[a-zA-Z0-9 _-]+$/)
