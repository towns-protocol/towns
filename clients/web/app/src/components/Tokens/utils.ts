import {
    Address,
    CheckOperationType,
    DecodedCheckOperation,
    IRuleEntitlementV2Base,
    NoopRuleData,
    createDecodedCheckOperationFromTree,
    createOperationsTree,
} from 'use-towns-client'
import { constants } from 'ethers'
import { formatUnits, parseUnits } from 'hooks/useBalance'
import { EVERYONE_ADDRESS } from 'utils'
import { TokenType } from './types'
import { Token, TokenEntitlement } from './TokenSelector/tokenSchemas'

export function convertTokenTypeToOperationType(type: TokenType) {
    switch (type) {
        case TokenType.ERC1155:
            return CheckOperationType.ERC1155 as const
        case TokenType.ERC721:
            return CheckOperationType.ERC721 as const
        case TokenType.ERC20:
            return CheckOperationType.ERC20 as const
        default:
            throw new Error('convertTokenTypeToOperationType: Invalid token type')
    }
}

export function convertOperationTypeToTokenType(type: CheckOperationType) {
    switch (type) {
        case CheckOperationType.ERC1155:
            return TokenType.ERC1155
        case CheckOperationType.ERC721:
            return TokenType.ERC721
        case CheckOperationType.ERC20:
            return TokenType.ERC20
        // Placeholder, needs fixing
        // https://linear.app/hnt-labs/issue/TOWNS-14181/handle-towns-that-are-gated-by-eth-balance
        case CheckOperationType.ETH_BALANCE:
            return TokenType.UNKNOWN
        default:
            console.error('convertOperationTypeToTokenType: Invalid operation type', type)
            return TokenType.UNKNOWN
    }
}

export function convertRuleDataToTokenEntitlementSchema(
    ruleData: IRuleEntitlementV2Base.RuleDataV2Struct,
): TokenEntitlement[] {
    return createDecodedCheckOperationFromTree(ruleData).map((p) => {
        const { threshold, tokenId, ...rest } = p
        return {
            ...rest,
            chainId: Number(p.chainId),
            type: convertOperationTypeToTokenType(p.type),
            quantity: threshold ?? 1n,
            tokenIds: tokenId ? [Number(tokenId)] : [],
        }
    })
}

export const transformQuantityForSubmit = (
    quantity: string,
    tokenType: TokenType,
    decimals: number = 18,
) => {
    return tokenType === TokenType.ERC20 ? parseUnits(quantity, decimals) : BigInt(quantity)
}

export const transformQuantityForDisplay = (
    quantity: bigint,
    tokenType: TokenType,
    decimals: number = 18,
) => {
    if (tokenType === TokenType.ERC20) {
        return formatUnits(quantity, decimals)
    } else {
        return quantity.toString()
    }
}

export const prepareGatedDataForSubmit = (
    gatingType: string,
    tokensGatedBy: Token[],
    usersGatedBy: string[],
    ethBalanceGatedBy: string,
) => {
    const finalUsersGatedBy = gatingType === 'everyone' ? [EVERYONE_ADDRESS] : usersGatedBy
    const finalTokensGatedBy = gatingType === 'everyone' ? [] : tokensGatedBy

    const tokensGatedByOps: DecodedCheckOperation[] = finalTokensGatedBy.map((t) => ({
        address: t.data.address as Address,
        chainId: BigInt(t.chainId),
        type: convertTokenTypeToOperationType(t.data.type),
        threshold: t.data.quantity
            ? transformQuantityForSubmit(t.data.quantity, t.data.type, t.data.decimals)
            : 1n,
        tokenId: t.data.tokenId ? BigInt(t.data.tokenId) : undefined,
    }))

    const ethBalanceOps: DecodedCheckOperation = {
        type: CheckOperationType.ETH_BALANCE,
        threshold: transformQuantityForSubmit(ethBalanceGatedBy, TokenType.ERC20, 18),
        chainId: 0n,
        address: constants.AddressZero,
    }

    const ruleData =
        finalTokensGatedBy.length > 0 || ethBalanceGatedBy
            ? createOperationsTree([
                  ...(finalTokensGatedBy.length > 0 ? tokensGatedByOps : []),
                  ...(ethBalanceGatedBy ? [ethBalanceOps] : []),
              ])
            : NoopRuleData

    return { tokensGatedBy: finalTokensGatedBy, usersGatedBy: finalUsersGatedBy, ruleData }
}
