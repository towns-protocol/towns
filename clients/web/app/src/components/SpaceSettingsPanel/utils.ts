import { TokenDataWithChainId } from '@components/Tokens/types'

type TokenDataStruct = {
    contractAddress: string
    chainId: number
    tokenIds: number[]
}

export function convertToNumber(val: string) {
    if (isNumeric(val)) {
        return parseFloat(val)
    } else {
        return NaN
    }
}

export function isNumeric(val: string) {
    return /^\d+(\.\d+)?$/.test(val)
}

const TOKEN_SPLITTER = '__TOKEN_ID__'

export function generateTokenIdKey(params: { contractAddress: string; tokenId: number }) {
    const { contractAddress, tokenId } = params
    return contractAddress + TOKEN_SPLITTER + tokenId
}

export function splitKeyToContractAddressAndTokenId(key: string) {
    return key.split(TOKEN_SPLITTER)
}

// Probably not needed w/ new xchain ux
export function mapTokenDataStructToTokenPillSelectorSelectionString(
    tokens: TokenDataStruct[],
): Set<string> {
    const selectedTokens = new Set<string>()
    // map tokens to a string like <address>__TOKEN_ID__<token_id>
    tokens.forEach((token: TokenDataStruct) => {
        if (token.tokenIds.length) {
            token.tokenIds.forEach((tokenId) => {
                selectedTokens.add(`${token.contractAddress}__TOKEN_ID__${tokenId}`)
            })
        } else {
            selectedTokens.add(token.contractAddress)
        }
    })
    return selectedTokens
}

// Probably not needed w/ new xchain ux
export function mapTokenPillSelectorSelectionStringToTokenDataStruct(tokenSelection: Set<string>) {
    // tokens is a set of strings like <address>__TOKEN_ID__<token_id> "0x1234__TOKEN_ID__1"
    // token_id is needed for ERC1155 tokens
    const tokenDataArray: TokenDataStruct[] = []
    tokenSelection.forEach((tokenIdString) => {
        const [contractAddress, tokenId] = splitKeyToContractAddressAndTokenId(tokenIdString)
        if (!contractAddress) {
            return
        }
        const match = tokenDataArray.find((x) => x.contractAddress === contractAddress)

        if (match) {
            if (tokenId !== undefined) {
                match.tokenIds.push(+tokenId)
            }
        } else {
            tokenDataArray.push({
                contractAddress,
                chainId: 1,
                tokenIds: tokenId !== undefined ? [+tokenId] : [],
            })
        }

        if (match?.tokenIds.length) {
            match.tokenIds = Array.from(new Set(match.tokenIds))
        }
    })
    return tokenDataArray
}

export function mapTokenOptionsToTokenDataStruct(tokens: TokenDataWithChainId[]) {
    return tokens.map((t) => {
        return {
            address: t.data.address,
            chainId: t.chainId,
            type: t.data.type,
            tokenIds: [],
            quantity: t.data.quantity ? BigInt(t.data.quantity) : 1n,
        }
    })
}
