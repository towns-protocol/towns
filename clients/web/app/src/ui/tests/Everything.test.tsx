import React from 'react'
import TestRenderer from 'react-test-renderer'
import { describe, expect, it } from 'vitest'
import { toJsonTree } from 'utils/test_util'
import { Everything } from './Everything'

describe('Everything', () => {
    it('should match previous snapshot', () => {
        const component = TestRenderer.create(<Everything />)
        const tree = toJsonTree(component)
        expect(tree).toMatchSnapshot()
    })
})
