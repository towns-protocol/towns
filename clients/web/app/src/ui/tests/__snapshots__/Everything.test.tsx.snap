// Vitest Snapshot v1

exports[`Everything > should match previous snapshot 1`] = `
<div
  className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4"
>
  <h1
    className="atoms_box__reo9080 textProperties_fontVariationSettings_strong_desktop__1dzo0x24 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_block_desktop__us3d3f0 textProperties_fontSize_xl32_desktop__1dzo0x2s textProperties_textTransform_none_desktop__1dzo0x21o textProperties_textAlign_left_desktop__1dzo0x210 Text__hecx1ow"
  >
    Heading
  </h1>
  <h1
    className="atoms_box__reo9080 textProperties_fontVariationSettings_strong_desktop__1dzo0x24 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_block_desktop__us3d3f0 textProperties_fontSize_xl32_desktop__1dzo0x2s textProperties_textTransform_none_desktop__1dzo0x21o textProperties_textAlign_left_desktop__1dzo0x210 Text__hecx1ow"
  >
    Heading 1
  </h1>
  <h2
    className="atoms_box__reo9080 textProperties_fontVariationSettings_strong_desktop__1dzo0x24 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_block_desktop__us3d3f0 textProperties_fontSize_xl24_desktop__1dzo0x2o textProperties_textTransform_none_desktop__1dzo0x21o textProperties_textAlign_left_desktop__1dzo0x210 Text__hecx1ow"
  >
    Heading 2
  </h2>
  <h3
    className="atoms_box__reo9080 textProperties_fontVariationSettings_strong_desktop__1dzo0x24 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_block_desktop__us3d3f0 textProperties_fontSize_xl_desktop__1dzo0x2k textProperties_textTransform_none_desktop__1dzo0x21o textProperties_textAlign_left_desktop__1dzo0x210 Text__hecx1ow"
  >
    Heading 3
  </h3>
  <h4
    className="atoms_box__reo9080 textProperties_fontVariationSettings_strong_desktop__1dzo0x24 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_block_desktop__us3d3f0 textProperties_fontSize_lg_desktop__1dzo0x2g textProperties_textTransform_none_desktop__1dzo0x21o textProperties_textAlign_left_desktop__1dzo0x210 Text__hecx1ow"
  >
    Heading 4
  </h4>
  <div
    className="atoms_box__reo9080 boxProperties_flexDirection_row_desktop__us3d3f16w boxProperties_display_flex_desktop__us3d3f4"
  >
    <div
      className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4"
    >
      H1
    </div>
    <div
      className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4"
    >
      H2
    </div>
  </div>
  <div
    className="atoms_box__reo9080 boxProperties_flexDirection_row_desktop__us3d3f16w boxProperties_display_flex_desktop__us3d3f4"
  >
    <div
      className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4"
    >
      H1
    </div>
    <div
      className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4"
    >
      H2
    </div>
  </div>
  <div
    className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4"
  >
    <div
      className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4 colorProperties_color_accent_lightMode__6xxol10"
    />
    <div
      className="atoms_box__reo9080 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_flex_desktop__us3d3f4 colorProperties_color_cta1_lightMode__6xxol1a"
    />
  </div>
  <span
    className="atoms_box__reo9080 textProperties_fontVariationSettings_normal_desktop__1dzo0x20 boxProperties_flexDirection_column_desktop__us3d3f170 boxProperties_display_block_desktop__us3d3f0 textProperties_fontSize_md_desktop__1dzo0x2c textProperties_textTransform_none_desktop__1dzo0x21o textProperties_textAlign_left_desktop__1dzo0x210 Text__hecx1ow"
  >
    Text
  </span>
  <svg
    fill="none"
    height="32"
    viewBox="0 0 200 67"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M71.8719 66.7008V2.51172H59.4158V66.7008H71.8719Z"
      fill="currentColor"
    />
    <path
      d="M155.499 66.7008V10.6484H157.207L159.417 25.5153L181.115 66.7008H200V2.51172H188.147V58.5642H186.439L184.128 43.6972L162.531 2.51172H143.646V66.7008H155.499Z"
      fill="currentColor"
    />
    <path
      d="M-5.18128e-07 14.3651L49.6235 14.3651L49.6235 16.0728L34.7566 25.8167L-2.0286e-06 48.9207L-2.8058e-06 66.7008L57.7602 66.7008L57.7602 55.9524L8.13666 55.9524L8.13666 54.2447L17.981 51.9343L57.7602 26.4194L57.7602 2.51172L0 2.51172L-5.18128e-07 14.3651Z"
      fill="currentColor"
    />
    <circle
      cx="107.232"
      cy="33.7018"
      fill="url(#paint0_linear_1205_31033)"
      r="33.7018"
    />
    <defs>
      <linearGradient
        gradientUnits="userSpaceOnUse"
        id="paint0_linear_1205_31033"
        x1="107.232"
        x2="107.232"
        y1="38.2975"
        y2="67.4037"
      >
        <stop
          stopColor="#C9356A"
        />
        <stop
          offset="0.364583"
          stopColor="#BB1B1B"
        />
        <stop
          offset="1"
          stopColor="#FA773F"
        />
      </linearGradient>
    </defs>
  </svg>
</div>
`;
